ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,467 @@
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    marker::PhantomData,
+    sync::{Arc, Mutex, Weak},
+};
+
+use async_trait::async_trait;
+use bitvec::{array::BitArray, prelude::Lsb0, slice::BitSlice};
+
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        context::{
+            dzkp_malicious::DZKPUpgraded as MaliciousDZKPUpgraded,
+            dzkp_semi_honest::DZKPUpgraded as SemiHonestDZKPUpgraded, Base, Context,
+            DZKPUpgradableContext, MaliciousContext, SemiHonestContext,
+        },
+        step::Gate,
+        RecordId,
+    },
+};
+
+/// `UnverifiedValues` are intermediate values that occur during a multiplication.
+/// These values need to be verified since there might have been malicious behavior.
+/// For a multiplication, let `(x1, x2), (x2, x3), (x3,x1)`, `(y1, y2), (y2, y3), (y3,y1)` be the shares of `x` and `y`
+/// and let the goal of the multiplication be to compute `x*y`.
+/// `(pi, p(i+1)` is the randomness from `PRSS` used for the multiplication by helper `i`.
+/// `z(i+1)` is the result of the multiplication received from the helper on the right.
+/// We do not need to store `zi` since it can be computed from the other stored values.
+#[derive(Clone, Debug)]
+struct UnverifiedValues {
+    x_left: BitArray<[u8; 256], Lsb0>,
+    x_right: BitArray<[u8; 256], Lsb0>,
+    y_left: BitArray<[u8; 256], Lsb0>,
+    y_right: BitArray<[u8; 256], Lsb0>,
+    prss_left: BitArray<[u8; 256], Lsb0>,
+    prss_right: BitArray<[u8; 256], Lsb0>,
+    z_right: BitArray<[u8; 256], Lsb0>,
+}
+
+impl UnverifiedValues {
+    fn new() -> Self {
+        Self {
+            x_left: BitArray::ZERO,
+            x_right: BitArray::ZERO,
+            y_left: BitArray::ZERO,
+            y_right: BitArray::ZERO,
+            prss_left: BitArray::ZERO,
+            prss_right: BitArray::ZERO,
+            z_right: BitArray::ZERO,
+        }
+    }
+
+    /// new from bitslices
+    /// ## Panics
+    /// Panics when length of slices is not 256
+    fn new_from_bitslices(
+        x_left: &BitSlice<u8, Lsb0>,
+        x_right: &BitSlice<u8, Lsb0>,
+        y_left: &BitSlice<u8, Lsb0>,
+        y_right: &BitSlice<u8, Lsb0>,
+        prss_left: &BitSlice<u8, Lsb0>,
+        prss_right: &BitSlice<u8, Lsb0>,
+        z_right: &BitSlice<u8, Lsb0>,
+    ) -> Self {
+        Self {
+            x_left: BitArray::try_from(x_left).unwrap(),
+            x_right: BitArray::try_from(x_right).unwrap(),
+            y_left: BitArray::try_from(y_left).unwrap(),
+            y_right: BitArray::try_from(y_right).unwrap(),
+            prss_left: BitArray::try_from(prss_left).unwrap(),
+            prss_right: BitArray::try_from(prss_right).unwrap(),
+            z_right: BitArray::try_from(z_right).unwrap(),
+        }
+    }
+}
+
+/// `Segment` is a variable size set or subset of `UnverifiedValues` using a `BitSlice`
+/// For efficiency reasons we constrain the segments size in bits to be a divisor of 256
```

Comment:
What is the efficiency reason? It seems like padding with `0`s is wasteful and not more efficient.

Reply:
The issue here is that we convert it into 256 bit arrays to have constant size elements. In the long run, we anyway prefer vectorizations so it would be a multiple of 256 bits. I discussed it with Andy and it made sense to us, to support non-vectorized operations, but not necessarily optimize for it.